kinvolk/azure-cvm-tooling
# vtpm-quote: Add SHA256 PCR value extraction API

This PR adds an API to extract all PCRs from SHA256 bank and store it in the quote. Alongwith this extraciton logic, it also adds a verification logic. Now users can call `quote.verify_pcrs()` to verify the PCR values.

## How to test

Here is a sample code that can be used to verify the workings of code. Run the following steps from the root of the directory:

```bash
mkdir -p az-cvm-vtpm/az-snp-vtpm/example2/src
```

```rust
cat <<EOF > az-cvm-vtpm/az-snp-vtpm/example2/src/main.rs
use az_snp_vtpm::vtpm;
use bincode;
use std::error::Error;
use std::fs;

fn main() -> Result<(), Box<dyn Error>> {
    // Save the PCRs into binary file.
    let nonce = "challenge".as_bytes().to_vec();
    let quote = vtpm::get_quote(&nonce)?;
    let quote_encoded: Vec<u8> = bincode::serialize(&quote).unwrap();
    fs::write("/tmp/quote.bin", quote_encoded).expect("Unable to write file");

    // Verification logic
    let quote_bytes = fs::read("/tmp/quote.bin").expect("Unable to read file");
    let quote: vtpm::Quote = bincode::deserialize(&quote_bytes[..]).unwrap();
    println!("Verify PCRs: {:?}", quote.verify_pcrs()?);
    Ok(())
}
EOF
```

Here is `Cargo.toml`

```toml
cat <<EOF > az-cvm-vtpm/az-snp-vtpm/example2/Cargo.toml
[package]
name = "snp-example2"
version = "0.1.0"
edition = "2021"

[dependencies]
az-snp-vtpm.path = "../"
bincode.workspace = true
openssl.workspace = true
EOF
```

Make the following change:

```diff
diff --git az-cvm-vtpm/Cargo.toml az-cvm-vtpm/Cargo.toml
index 4a36f15..4eec2e9 100644
--- az-cvm-vtpm/Cargo.toml
+++ az-cvm-vtpm/Cargo.toml
@@ -13,6 +13,7 @@ members = [
        "az-snp-vtpm",
        "az-tdx-vtpm",
        "az-snp-vtpm/example",
+       "az-snp-vtpm/example2",
 ]

 [lib]
```

Finally run the code:

```bash
cd az-cvm-vtpm/az-snp-vtpm/example2
cargo build && sudo ../../target/debug/snp-example2
```
